opal_client/data/updater.py

```diff
@@ -233,5 +203,83 @@ async def wait_until_done(self):
+
+            logger.info("Reporting the update to requested callbacks", urls=urls)
+            await data_fetcher.handle_urls(urls)
+            # TODO: log failed reports
+        except:
+            logger.exception("Failed to excute report_update_results")
+
+    async def update_policy_data(self, update: DataUpdate = None, policy_store: BasePolicyStoreClient = None, data_fetcher=None):
+        """
+        fetches policy data (policy configuration) from backend and updates it into policy-store (i.e. OPA)
+        """
+        policy_store = policy_store or DEFAULT_POLICY_STORE_GETTER()
+        if data_fetcher is None:
+            data_fetcher = DataFetcher()
+        # types / defaults
+        urls: List[Tuple[str, FetcherConfig]] = None
+        entries: List[DataSourceEntry] = []
+        # track the result of each url in order to report back
+        reports: List[DataEntryReport] = []
+        # if we have an actual specification for the update
+        if update is not None:
+            entries = update.entries
+            urls = [(entry.url, entry.config) for entry in entries]
 
+        # get the data for the update
+        logger.info("Fetching policy data", urls=urls)
+        # Urls may be None - handle_urls has a default for None
+        policy_data_with_urls = await data_fetcher.handle_urls(urls)
+        # Save the data from the update
+        # We wrap our interaction with the policy store with a transaction  
+        async with policy_store:
+            for (url, fetch_config, result), entry in itertools.zip_longest(policy_data_with_urls, entries):
+                if not isinstance(result, Exception):
+                    # get path to store the URL data (default mode (None) is as "" - i.e. as all the data at root)
+                    policy_store_path = "" if entry is None else entry.dst_path
+                    # None is not valid - use "" (protect from missconfig)
+                    if policy_store_path is None:
+                        policy_store_path = ""
+                    # fix opa_path (if not empty must start with "/" to be nested under data)
+                    if policy_store_path != "" and not policy_store_path.startswith("/"):
+                        policy_store_path = f"/{policy_store_path}"
+                    policy_data = result
+                    # Create a report on the data-fetching
+                    report = DataEntryReport(entry=entry, hash=self.calc_hash(policy_data), fetched=True)
+                    logger.info(
+                        "Saving fetched data to policy-store: source url='{url}', destination path='{path}'",
+                        url=url,
+                        path=policy_store_path or '/'
+                    )
+                    res = await policy_store.set_policy_data(policy_data, path=policy_store_path)
+                    # a valuse of None indicates a failure to save to the policy store
+                    report.saved = res is not None
```

Comment:
won't work, i risk repeating my comment but:
- if OPA call is failed, we either get a bad status code or some exception which we re-raise.
- 'None' content is actually valid since if successful OPA returns 204 No content
- i think instead of expecting None you should try catch and mark saved if there is no exception raised.

Reply:
I can't try / catch since set_policy wrapped with fail_silently().
Was trying to avoid changing that part.
Suggestions?

Reply:
i guess we failed silently before because we didn't want a single failed write to cause problems.
however now that we have success reports, maybe we should remove this behavior.

options i see, both are ok as far as i am concerned:
1) remove fail silently - will require us to put try-catch in all the call sites and handle it there, but i am not sure this is very bad.
2) do what you suggested with None as failure marker -> will require changing the existing implementation of OPA client to return a fake success value, something like {'status': 'ok'} when opa returns 204.

WDYT?

Reply:
I'll remove the fail silently